keycloak-event-emitter/src/main/java/io/cloudtrust/keycloak/kafkaeventemitter/KafkaEventEmitterProvider.java

```diff
@@ -0,0 +1,129 @@
+    }
+
+    @Override
+    public void onEvent(Event event) {
+        completeEventAttributes(event);
+        long uid = idGenerator.nextValidId();
+        IdentifiedEvent identifiedEvent = new IdentifiedEvent(uid, event);
+
+        // Flatbuffer serialization
+        ByteBuffer buffer = SerialisationUtils.toFlat(identifiedEvent);
+
+        produceEvent(buffer, eventTopic);
+    }
+
+    @Override
+    public void onEvent(AdminEvent adminEvent, boolean includeRepresentation) {
+        long uid = idGenerator.nextValidId();
+        IdentifiedAdminEvent identifiedAdminEvent = new IdentifiedAdminEvent(uid, adminEvent);
+        ExtendedAdminEvent customAdminEvent = completeAdminEventAttributes(identifiedAdminEvent);
+
+        // Flatbuffer serialization
+        ByteBuffer buffer = SerialisationUtils.toFlat(customAdminEvent);
+
+        produceEvent(buffer, adminEventTopic);
+    }
+
+    @Override
+    public void close() {
+
+    }
+
+    private void completeEventAttributes(Event event) {
+        // add username if missing
+        if (event.getDetails() == null) {
+            event.setDetails(new HashMap<>());
+        }
+        String eventUsername = event.getDetails().get(Details.USERNAME);
+        if (!Strings.isNullOrEmpty(event.getUserId()) && Strings.isNullOrEmpty(eventUsername)) {
+            RealmModel realm = keycloakSession.realms().getRealm(event.getRealmId());
+            // retrieve username from userId
+            UserModel user = keycloakSession.users().getUserById(realm, event.getUserId());
+            if (user != null) {
+                event.getDetails().put(Details.USERNAME, user.getUsername());
+            }
+        }
+    }
+
+    private ExtendedAdminEvent completeAdminEventAttributes(IdentifiedAdminEvent adminEvent) {
+        ExtendedAdminEvent extendedAdminEvent = new ExtendedAdminEvent(adminEvent);
+        // add always missing agent username
+        ExtendedAuthDetails extendedAuthDetails = extendedAdminEvent.getAuthDetails();
+        if (!Strings.isNullOrEmpty(extendedAuthDetails.getUserId())) {
+            RealmModel realm = keycloakSession.realms().getRealm(extendedAuthDetails.getRealmId());
+            UserModel user = keycloakSession.users().getUserById(realm, extendedAuthDetails.getUserId());
+            extendedAuthDetails.setUsername(user.getUsername());
+        }
+        // add username if resource is a user
+        String resourcePath = extendedAdminEvent.getResourcePath();
+        if (resourcePath != null && resourcePath.startsWith("users")) {
+            // parse userID
+            String pattern = "^users/([0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12})$";
+            Pattern r = Pattern.compile(pattern);
+            Matcher m = r.matcher(resourcePath);
+            if (m.matches()) {
+                String userId = m.group(1);
+                RealmModel realm = keycloakSession.realms().getRealm(adminEvent.getRealmId());
```

Comment:
Code duplicated... may be you could use something like this:
```
this.findUser(userId, u -> extendedAdminEvent.getDetails().put("username", user.getUsername()));
.
private void findUser(String userID, Consumer<UserModel> whenUserFound) {
    RealmModel realm = ...
    if (realm!=null) {
        UserModel user = ...
        if (user!=null) {
            whenUserFound.accept(user);
        }
    }
}
`

Reply:
done